src/app/services/courses-store.service.ts

```diff
@@ -1,42 +1,167 @@
-        // Add your code here
-    }
+  getAll(): Observable<Course[]> {
+    // Add your code here
+    this.isLoading$$.next(true);
+    return this.coursesService.getAll().pipe(
+      switchMap((res) => {
+        return forkJoin(
+          res.result.map((course: any) =>
+            forkJoin(
+              course.authors.map((authorId: string) =>
+                this.getAuthorById(authorId)
+              )
+            ).pipe(map((authors) => ({ ...course, authors })))
+          )
+        );
+      }),
+      tap((fixedCourses: any) => {
+        this.courses$$.next(fixedCourses);
+        this.isLoading$$.next(false);
+      }),
+      catchError((error) => {
+        this.handleError(error);
+        this.courses$$.next([]);
+        return of([]);
+      })
+    );
+  }
 
-    deleteCourse(id: string) {
-        // Add your code here
-    }
+  createCourse(course: CourseCreate): void {
+    // replace 'any' with the required interface
+    // Add your code here
+    this.coursesService.createCourse(course).subscribe((res) => {
+      console.log(res);
+      this.router.navigate(["/courses"]);
+      this.getAll();
+    });
+  }
 
-    filterCourses(value: string) {
-        // Add your code here
-    }
+  getCourse(id: string): Observable<Course> {
+    // Add your code here
+    this.isLoading$$.next(true);
+    return this.coursesService.getCourse(id).pipe(
+      switchMap((res) => {
+        const course = res.result;
+        return forkJoin(
+          course.authors.map((authorId: string) =>
+            this.getAuthorById(authorId).pipe(
+              map((authorName) => ({ id: authorId, name: authorName }))
+            )
+          )
+        ).pipe(map((authors) => ({ ...course, authors })));
+      })
+    );
+  }
 
-    getAllAuthors() {
-        // Add your code here
-    }
+  editCourse(id: string, course: CourseCreate): void {
+    // replace 'any' with the required interface
+    // Add your code here
+    this.isLoading$$.next(true);
+    this.coursesService.editCourse(id, course).subscribe(() => this.getAll());
+  }
 
-    createAuthor(name: string) {
-        // Add your code here
-    }
+  deleteCourse(id: string): void {
+    // Add your code here
+    this.isLoading$$.next(true);
+    this.coursesService.deleteCourse(id).subscribe(() => this.getAll());
+  }
+
+  filterCourses(value: string): Observable<Course[]> {
+    // Add your code here
+    this.isLoading$$.next(true);
+    return this.coursesService.filterCourses(value).pipe(
+      switchMap((res) => {
+        if (res.result.length === 0) {
+          return of([]);
+        }
+        return forkJoin(
+          res.result.map((course: any) =>
+            forkJoin(
+              course.authors.map((authorId: string) =>
+                this.getAuthorById(authorId)
+              )
+            ).pipe(map((authors) => ({ ...course, authors })))
+          )
+        );
+      }),
```

Comment:
Hopefully, you will recall what is happening here, after 1 month merging the PR. The first two points we could enhance here are to make the nesting more flat and optimize requests by missing duplicated authors.